src/p_mobj.c

```diff
@@ -110,6 +110,40 @@ boolean P_SetMobjState(mobj_t* mobj,statenum_t state)
   return ret;
 }
 
+// [crispy] return the latest "safe" state in a state sequence,
+// so that no action pointer is ever called
+static statenum_t P_LatestSafeState(statenum_t state)
+{
+  statenum_t safestate = S_NULL;
+  static statenum_t laststate, lastsafestate;
+
+  if (state == laststate)
+  {
+    return lastsafestate;
+  }
+
+  for (laststate = state; state != S_NULL; state = states[state].nextstate)
```

Comment:
This could be an infinite loop, right? 🤔

Reply:
Even if it goes through all states, and even if state is never S_NULL, there is a check in a for loop state == states[state].nextstate that will break it. Or are you referring to something else?

Reply:
What if there's a state cycle a to b to c to a, and all of the states have action pointers?

Reply:
Ok, I see what you mean.

Basically, we are calling this function only when we are trying to spawn bullet puffs (the state is always S_PUFF1). The loop will always cycle through S_PUFF1, S_PUFF2, S_PUFF3 and S_PUFF4. After that the state will be equal to 0.

AFAIU it can't go from S_PUFF3 to S_PUFF1? Or can it?

Reply:
I'm talking about dehacked modifications. We can't assume how things will be configured - but maybe this is an accepted assumption in woof. Just pointing it out.

Reply:
> Basically, we are calling this function only when we are trying to spawn bullet puffs (the state is always S_PUFF1). 

Not only that. At least in Crispy it is also called for missiles to show at least some frames of an explosion. 

> I'm talking about dehacked modifications. We can't assume how things will be configured - but maybe this is an accepted assumption in woof. Just pointing it out.

The code has worked as is in Crispy for about six years now. But maybe you are right anyway. Today I would be more careful than I was back then. 

There is really only one way to detect a complex state cycle and that is to count the number of iterations and error out of an arbitrary limit is reached, e.g.
https://github.com/fabiangreffrath/crispy-doom/blob/68196ab364cae3b97349dd9fe5e08e1184bd05ce/src/doom/p_mobj.c#L86

Reply:
In this case you could track what states you've seen and abort if you hit the same one twice. That's also how the real state updater works in pr+ / dsda-doom. I also think the risk is higher in woof than in crispy because the higher complevels tend to be used for more complex projects. Still, I don't think it's necessarily super likely.